In [1]:
import pandas as pd
import numpy as np

In [2]:
from sklearn.datasets import load_boston

In [25]:
boston = load_boston()

In [4]:
type(boston)

sklearn.utils.Bunch

In [5]:
boston.keys()

dict_keys(['data', 'target', 'feature_names', 'DESCR', 'filename', 'data_module'])

In [6]:
boston_df = boston.data

In [7]:
boston_df.shape

(506, 13)

In [8]:
target = boston["target"]
feature_names = boston["feature_names"]

In [9]:
df = pd.DataFrame(boston_df, columns=feature_names)

In [10]:
type(target)

numpy.ndarray

In [11]:
y_target = pd.DataFrame(target, columns=['price'])

In [12]:
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 13 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    float64
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    float64
 9   TAX      506 non-null    float64
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
dtypes: float64(13)
memory usage: 51.5 KB


CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03  
3     18.7  394.63   2.94  
4     18.7  396.90   5.33

In [13]:
y_target.info()
y_target.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   price   506 non-null    float64
dtypes: float64(1)
memory usage: 4.1 KB


price
0   24.0
1   21.6
2   34.7
3   33.4
4   36.2

---

In [14]:
from sklearn.model_selection import train_test_split

In [15]:
x_train, x_test, y_train, y_test = train_test_split(df, y_target, test_size=0.3, random_state=42)

---

In [16]:
from sklearn.linear_model import LinearRegression

In [17]:
lr = LinearRegression()

In [18]:
lr.fit(x_train, y_train);

In [19]:
y_pred = lr.predict(x_test)

In [30]:
y_pred

array([[28.64896005],
       [36.49501384],
       [15.4111932 ],
       [25.40321303],
       [18.85527988],
       [23.14668944],
       [17.3921241 ],
       [14.07859899],
       [23.03692679],
       [20.59943345],
       [24.82286159],
       [18.53057049],
       [-6.86543527],
       [21.80172334],
       [19.22571177],
       [26.19191985],
       [20.27733882],
       [ 5.61596432],
       [40.44887974],
       [17.57695918],
       [27.44319095],
       [30.1715964 ],
       [10.94055823],
       [24.02083139],
       [18.07693812],
       [15.934748  ],
       [23.12614028],
       [14.56052142],
       [22.33482544],
       [19.3257627 ],
       [22.16564973],
       [25.19476081],
       [25.31372473],
       [18.51345025],
       [16.6223286 ],
       [17.50268505],
       [30.94992991],
       [20.19201752],
       [23.90440431],
       [24.86975466],
       [13.93767876],
       [31.82504715],
       [42.56978796],
       [17.62323805],
       [27.01963242],
       [17

In [20]:
from sklearn.metrics import r2_score

In [21]:
R2 = r2_score(y_test, y_pred)

##### ОТВЕТ

R^2 = 0.7112260057484974

---
TASK 2
---

In [22]:
from sklearn.ensemble import RandomForestRegressor

In [26]:
model = RandomForestRegressor(n_estimators=100, max_depth=12, random_state=42)

In [27]:
model.fit(x_train, y_train.values[:, 0])

RandomForestRegressor(max_depth=12, random_state=42)

In [28]:
y_pred_rf = model.predict(x_test)

In [29]:
y_pred_rf

array([22.99461365, 31.165     , 15.9295    , 24.26660838, 16.98754167,
       21.77113098, 19.80115808, 15.04516667, 21.46599671, 21.21497951,
       19.88368409, 20.34809224,  7.962     , 21.29828287, 19.29768133,
       25.02238561, 18.71827273,  8.385     , 46.011     , 15.05353611,
       24.28334244, 24.02785755, 14.70456667, 23.45338622, 14.79495833,
       15.28944167, 21.97226809, 13.66391667, 19.71284667, 21.00210276,
       19.68960595, 23.38091772, 30.66275   , 20.21730373, 14.32159048,
       16.17276667, 33.56832143, 18.96360565, 20.870242  , 23.89041728,
       18.06433333, 28.9582    , 45.587     , 19.72523679, 22.71732832,
       14.243025  , 15.85610833, 24.25625848, 19.06644312, 28.44033333,
       21.43756635, 34.06649134, 17.95774167, 25.8496381 , 45.37971429,
       21.54203521, 15.78007262, 31.802     , 22.25318349, 20.45276667,
       24.92604697, 33.71109848, 29.25896429, 19.14772225, 27.44115   ,
       16.30433333, 13.90119167, 23.0693973 , 28.8535    , 15.23

In [33]:
R2_rf =r2_score(y_test, y_pred_rf)

In [34]:
R2_rf

0.8713629100115731

# ОТВЕТ

r 2 на основе решения случайного леса ближе к 1

0.8713629100115731 >  0.7112260057484974

###### Следовательно, модель случайного леса работает лучше линейной регрессии
